### PR TITLE
Update Actions and Repo, and Add New Actions

### PR DESCRIPTION
**Pull request by erogluorhan**
NOTE: Huge number of files changed is due to `pre-commit hook` that is newly setup and ran on all the files in the repo.

This PR is aimed at making this repo look as much applicable/possible as GeoCAT-comp repo with respect to actions, folder structure, etc. for consistency purposes. Therefore it makes the following changes:

- Please first beware that the Github actions "build_test" is now renamed to "CI/Python (...)". Thus, for review/approval, see if the newly named actions are passing even though the older name is required by branch protection rules.

- Adds automated `ci.yml` and removes manually-triggered `build_test.yml`.

- Renames CI action and step names similar to geocat-comp wherever applicable.

- Adds `pre-commit hooks` and removes 'yapf.yaml':
Note: To use these pre-commit hooks in local, each developer needs to do the followings (after pulling this branch or `main` branch after merge into local):
   `pip install pre-commit`
   `pre-commit install`

- Adds a LICENSE file and then several badges into README.md to have similar badges with GeoCAT-comp.

- Adds code coverage reports, `codecov`, as in GeoCAT-comp.

- Adds an action for uploading GeoCAT-f2py tp PyPi.

### CODE REVIEW

**Comment by erogluorhan** [comment]
Hey @andersy005 , I am adding code coverage to this repo, too. Reports are being generated correctly, I also activated codecov webhooks; however, there is still an issue with"enabling the GeoCAt-f2py repo" in my codecov account. Here is the message:

image

The "installation settings" link there is this URL: https://github.com/organizations/NCAR/settings/installations/591816

I think I don't have an Admin status in NCAR organization, and it may be the reason I cannot see above URL and access to NCAR' s organization settings. 

Do you have an idea how to enable GeoCAT-f2py as you did previously for GeoCAT-comp?

**Comment by andersy005** [comment]
> Do you have an idea how to enable GeoCAT-f2py as you did previously for GeoCAT-comp? 

Since this repository is public, I don't think you have to do anything else. I can confirm that the reports are publicly accessible: https://codecov.io/gh/NCAR/geocat-f2py/tree/a987975b7b69532dbc050607a1b1c08ecbe8ee8a 

So, I'd say that everything is working as expected.

File: .github/codecov.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        threshold: 0.2
+        if_not_found: success
+    patch:
+      default:
+        enabled: no
+        if_not_found: success
+    changes:
+      default:
+        enabled: no
+        if_not_found: success
```
Comment (andersy005) [inline .github/codecov.yml:21]: This file shouldn't be here. It should be in the root directory.

Comment (erogluorhan) [reply .github/codecov.yml:21]: > This file shouldn't be here. It should be in the root directory.

I don't think that's an issue since (1) [codecov docs](https://docs.codecov.io/docs/codecov-yaml) do not explicitly say it should be in the root directory, (2) GeoCAT-comp also has that file under `.github` folder and its badge works fine:

image


But, when you check README of this branch here (`update_actions`), it cannot show the report outputs:

image

I am checking the codecov badge and URL, and they seem like good, too.

Comment (erogluorhan) [reply .github/codecov.yml:21]: Ok, I now realize there is nothing to be concerned here since the same warning (i.e. "repo is not enabled") is also being given for GeoCAT-comp but its overview is correctly populated from `main` branch. Once this PR is merged, the badge here as well as the overview at codecov should be correctly populated. 

Thanks for discussing!

**Comment by pilotchute** [comment]
Was preparing to give my review, when I saw that some automated function had modified the fortran files. While I do not expect that any of the fortran files have been broken, I do not like that they have been touched at all. They need to remain untouched, in order to preserve our confidence that they do in fact provide the exact desired behavior from NCL.

**Comment by erogluorhan** [comment]
> Was preparing to give my review, when I saw that some automated function had modified the fortran files. While I do not expect that any of the fortran files have been broken, I do not like that they have been touched at all. They need to remain untouched, in order to preserve our confidence that they do in fact provide the exact desired behavior from NCL.

This is a great catch, thanks! I totally agree with this.

It was the `pre-commit hooks` that were run on all the files and I missed the files under `src/geocat/f2py/fortran` were also changed. 

I fixed it now by reverting all of those files as well as adding an `exclude` key to all of our pre-commit hooks for this repo(see [.pre-commit-config.yaml)](https://github.com/NCAR/geocat-f2py/blob/update_actions/.pre-commit-config.yaml) as follows:

`exclude: src/geocat/f2py/fortran/`

I need your input here: With this exclusion, we are ignoring not only Fortran files but also all the files such as `.pyf` we implemented under `src/geocat/f2py/fortran/` directory. My thought is that we can/should only exclude Fortran files. What do you think?

**Comment by pilotchute** [comment]
`.pyf` files are not a python file type, and the linter shouldn't touch them. I am happy to exclude all of the files in `.../fortran/` since those are only used for compilation during the build phase, and they should remain untouched once tested and approved.

**Comment by erogluorhan** [comment]
> `.pyf` files are not a python file type, and the linter shouldn't touch them. I am happy to exclude all of the files in `.../fortran/` since those are only used for compilation during the build phase, and they should remain untouched once tested and approved.

Agreed!

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,37 @@
+repos:
+-   repo: https://github.com/pre-commit/mirrors-yapf    # To format the code to conform YAPF
+    rev: v0.30.0
+    hooks:
+    - id: yapf
+      args: ['--style=.style.yapf', '--recursive', '--in-place']
```
Comment (pilotchute) [inline .pre-commit-config.yaml:6]: This is a different style than we've been using, I would suggest remaining on the google style.

Comment (pilotchute) [reply .pre-commit-config.yaml:6]: Nevermind, this is referncing a style file which is pulling google style. all good

File: fast_test.py [hunk 3]
```diff
@@ -1,12 +1,14 @@
 # THIS IS A DEVELOPMENT PERFORMANCE TESTING CLASS, REMOVE WHEN NO LONGER NEEDED
```
Comment (pilotchute) [inline fast_test.py:1]: This file should not be included in the commit, it isn't an official part of the repo, I don't know how it got into main.

File: setup.py [hunk 4]
```diff
@@ -15,16 +21,42 @@ def version():
...
+      maintainer_email='geocat@ucar.edu',
+      python_requires='>=3.6',
+      install_requires=requirements,
+      description="""GeoCAT-f2py wraps, in Python, the compiled language
+    implementations of some of the computational functions found under the
+    GeoCAT-comp umbrella.""",
+      long_description=long_description,
+      long_description_content_type='text/markdown',
+      classifiers=[
+          'Operating System :: OS Independent',
```
Comment (pilotchute) [inline setup.py:36]: Not true, we are not OS independent for f2py, at least yet. We don't have a build process for windows.

File: src/geocat/f2py/__init__.py [hunk 5]
```diff
@@ -1,10 +1,12 @@
...
+from .dpres_plevel_wrapper import dpres_plevel
 from .errors import *
-from .linint2_wrapper import (linint1, linint2, linint2pts, linint2_points)
-from .missing_values import (py2fort_msg, fort2py_msg)
-from .moc_globe_atl_wrapper import (moc_globe_atl)
-from .rcm2points_wrapper import (rcm2points)
-from .rcm2rgrid_wrapper import (rcm2rgrid, rgrid2rcm)
-from .triple_to_grid_wrapper import (grid_to_triple, triple_to_grid,
-                                     grid2triple, triple2grid)
+from .linint2_wrapper import linint1, linint2, linint2_points, linint2pts
```
Comment (pilotchute) [inline src/geocat/f2py/__init__.py:6]: What is removing the grouping parenthesis? From a style perspective they are quite useful as they allow longer import lists to be multiline.

File: src/geocat/f2py/checks.py [hunk 6]
```diff
@@ -1,6 +1,6 @@
-import xarray as xr
-import numpy as np
 import dask as da
+import numpy as np
+import xarray as xr
```
Comment (pilotchute) [inline src/geocat/f2py/checks.py:3]: What is reordering the imports?

File: src/geocat/f2py/rcm2points_wrapper.py [hunk 7]
```diff
@@ -75,19 +79,34 @@ def rcm2points(lat2d, lon2d, fi, lat1d, lon1d, opt=0, msg=None, meta=False):
...
     Description
     -----------
 
+<<<<<<< HEAD
+        Interpolates data on a curvilinear grid, such as those used by the RCM (Regional Climate Model),
+        WRF (Weather Research and Forecasting) and NARR (North American Regional Reanalysis)
+        models/datasets to an unstructured grid. All of these have latitudes that are oriented south-to-north.
+        A inverse distance squared algorithm is used to perform the interpolation.
+        Missing values are allowed and no extrapolation is performed.
+=======
```
Comment (pilotchute) [inline src/geocat/f2py/rcm2points_wrapper.py:103]: This looks like a merge conflict got stuck into the docstring.

File: src/geocat/f2py/rcm2rgrid_wrapper.py [hunk 8]
```diff
@@ -164,10 +163,10 @@ def rcm2rgrid(lat2d, lon2d, fi, lat1d, lon1d, msg=None, meta=False):
...
         fi = xr.DataArray(
             fi.data,
-            # coords={
-            #     fi.dims[-1]: lon2d,
-            #     fi.dims[-2]: lat2d,
-            # },
+        # coords={
+        #     fi.dims[-1]: lon2d,
+        #     fi.dims[-2]: lat2d,
+        # },
```
Comment (pilotchute) [inline src/geocat/f2py/rcm2rgrid_wrapper.py:169]: I disagree with this action by the linter.

File: src/geocat/f2py/triple_to_grid_wrapper.py [hunk 9]
```diff
@@ -361,10 +361,10 @@ def triple_to_grid(data,
...
         data = xr.DataArray(
             data.data,
-            # coords={
-            #     data.dims[-1]: x_in,
-            #     data.dims[-2]: y_in,
-            # },
+        # coords={
+        #     data.dims[-1]: x_in,
+        #     data.dims[-2]: y_in,
+        # },
```
Comment (pilotchute) [inline src/geocat/f2py/triple_to_grid_wrapper.py:367]: another linter modification I don't agree with.
Why is Yapf behaving differently with this implementation? it should still be using the google style that we already had right?

File: fast_test.py (hunk 3 above)
Comment (erogluorhan) [reply fast_test.py]: Removed.

File: test/test_moc_globe_atl.py [hunk 10]
```diff
@@ -65,70 +72,64 @@ class Test_Moc_Globe_Atl(ut.TestCase):
...
-        out_arr = geocat.f2py.moc_globe_atl(
-            xr.DataArray(lat_aux_grid), xr.DataArray(tmp_a.astype(np.float64)),
-            xr.DataArray(tmp_a.astype(np.float64)),
-            xr.DataArray(tmp_a.astype(np.float64)), xr.DataArray(t_lat),
-            xr.DataArray(rmlak))
+        out_arr = moc_globe_atl(xr.DataArray(lat_aux_grid),
+                                xr.DataArray(tmp_a.astype(np.float64)),
+                                xr.DataArray(tmp_a.astype(np.float64)),
+                                xr.DataArray(tmp_a.astype(np.float64)),
+                                xr.DataArray(t_lat), xr.DataArray(rmlak))
```
Comment (pilotchute) [inline test/test_moc_globe_atl.py:88]: Another linter change I don't like. Prev version was cleaner and more compact.

File: setup.py (hunk 4 above)
Comment (erogluorhan) [reply setup.py]: Great catch (of stealing from GeoCAT-comp :) ). Corrected to MacOS X and Linux now.

File: test/test_rcm2points.py [hunk 11]
```diff
@@ -68,131 +74,97 @@
...
         nt.assert_array_almost_equal(
             fo_nom_opt2_expected,
-            geocat.f2py.rcm2points(lat2d,
-                                   lon2d,
-                                   fi_nom.astype(np.float64),
-                                   lat,
-                                   lon,
-                                   opt=2))
+            rcm2points(lat2d, lon2d, fi_nom.astype(np.float64), lat, lon,
+                       opt=2))
```
Comment (pilotchute) [inline test/test_rcm2points.py:90]: I do like this linter change. We should talk about all the differences we're seeing between the two styles that should have been the same that we are using.

**Review by pilotchute — COMMENTED** [review]
Files reviewed. I think we still have a conversation ahead about some of the automated changes.

File: src/geocat/f2py/checks.py (hunk 6 above)
Comment (erogluorhan) [reply src/geocat/f2py/checks.py:3]: It is the pre-commit hook `isort` (i.e. "import sort"). See inside [.pre-commit-config.yaml](https://github.com/NCAR/geocat-f2py/blob/update_actions/.pre-commit-config.yaml) 

```
-   repo: https://github.com/PyCQA/isort    # To sort the imports
    rev: 5.7.0
    hooks:
    - id: isort
      exclude: src/geocat/f2py/fortran/
```

File: src/geocat/f2py/rcm2points_wrapper.py (hunk 7 above)
Comment (erogluorhan) [reply src/geocat/f2py/rcm2points_wrapper.py]: Fixed, thanks!

File: src/geocat/f2py/rcm2rgrid_wrapper.py (hunk 8 above)
Comment (erogluorhan) [reply src/geocat/f2py/rcm2rgrid_wrapper.py]: I agree! It was due to a strange behavior of `yapf`, and I fixed it now. FYI: Below is why there was such a behavior:

When we were handling the Google style for `yapf` from inside the style file (`based_on_style = google` in `.style.yapf` in the root directory), there was that weird behavior you pointed out. When I removed that setting from style file and instead put it into `.pre-commit-config.yaml` as an argument of `yapf`  (i.e. `args: ['--style=.style.yapf', '--in-place', '--recursive', '--style', 'google']`), it fixed the thing.

I am still keeping the `yapf` style file, because even though there are a few style configs in it for now, there are many [YAPF Knobs](https://github.com/google/yapf#knobs) that we would be useful in the future.

File: src/geocat/f2py/triple_to_grid_wrapper.py (hunk 9 above)
Comment (erogluorhan) [reply src/geocat/f2py/triple_to_grid_wrapper.py]: See above resolution.

File: src/geocat/f2py/rcm2rgrid_wrapper.py (hunk 8 above)
Comment (pilotchute) [reply src/geocat/f2py/rcm2rgrid_wrapper.py]: Thanks that's great.

File: src/geocat/f2py/triple_to_grid_wrapper.py (hunk 9 above)
Comment (pilotchute) [reply src/geocat/f2py/triple_to_grid_wrapper.py]: resolved

File: test/test_moc_globe_atl.py (hunk 10 above)
Comment (pilotchute) [reply test/test_moc_globe_atl.py:88]: OBE

File: test/test_rcm2points.py (hunk 11 above)
Comment (pilotchute) [reply test/test_rcm2points.py:90]: OBE

File: src/geocat/f2py/checks.py (hunk 6 above)
Comment (pilotchute) [reply src/geocat/f2py/checks.py:3]: thanks resolveing

File: test/test_moc_globe_atl.py (hunk 10 above)
Comment (erogluorhan) [reply test/test_moc_globe_atl.py:88]: Here is how it looked like in `main` branch:

```
        out_arr = geocat.f2py.moc_globe_atl(
            xr.DataArray(lat_aux_grid), xr.DataArray(tmp_a.astype(np.float64)),
            xr.DataArray(tmp_a.astype(np.float64)),
            xr.DataArray(tmp_a.astype(np.float64)), xr.DataArray(t_lat),
            xr.DataArray(rmlak))
```

I don't agree that it was cleaner and compact because it did not look consistent as the rest of the code (e.g. Lines 90-94) had the function arguments just aligned to each other:

```
out_arr = geocat.f2py.moc_globe_atl(lat_aux_grid,
                                    tmp_a.astype(np.float32),
                                    tmp_a.astype(np.float32),
                                    tmp_a.astype(np.float32), t_lat,
                                    rmlak)
```

This is more of a personal preferences difference though, and I could be fine with either way.

File: src/geocat/f2py/__init__.py (hunk 5 above)
Comment (erogluorhan) [reply src/geocat/f2py/__init__.py:6]: It is again pre-commit hook `isort` (i.e. "import sort"). Actually there is a nice behavior of it on these kind of imports  which I really like and would like to share with you:

When it finds multiple imports in a single line where number of chars exceed 79 (or whatever), it puts automatic parentheses and split that line into two or more lines. However, if you just put parentheses into where multiple imports can already fit into a single line, it then removes parentheses. In other words, parentheses are only used for cases where imports do not fit into a single line, which I like. 

Thoughts?

**Comment by erogluorhan** [comment]
> Files reviewed. I think we still have a conversation ahead about some of the automated changes.

Thanks for your through review!

I have addressed all of your comments, so I just requested your re-review. Though, I am happy to have a quick chat to talk about any unclear points.

File: src/geocat/f2py/__init__.py (hunk 5 above)
Comment (pilotchute) [reply src/geocat/f2py/__init__.py:6]: That's fine, I don't like the un-needed modifications (which I think hurt readability), but I'll not worry about it as long as multiline imports won't be broken.

**Comment by pilotchute** [comment]
> > Files reviewed. I think we still have a conversation ahead about some of the automated changes.
> 
> Thanks for your thorough review!
> 
> I have addressed all of your comments, so I just requested your re-review. Though, I am happy to have a quick chat to talk about any unclear points.

Thanks, I'll have my review done soon.

**Comment by erogluorhan** [comment]
@pilotchute  FYI: I removed style file now since it was almost empty with its current shape and there was no need to keep it in repo. If/when we need to employ more yapf knobs, we can add again. If you think otherwise, let me know.

File: test/test_rcm2points.py (hunk 11 above)
Comment (erogluorhan) [reply test/test_rcm2points.py:90]: > I do like this linter change. We should talk about all the differences we're seeing between the two styles that should have been the same that we are using.

Now, I spent more time on it, and here are my findings:

There is no difference between the behaviors of our older YAPF config or that in this PR. The main difference is that we called `geocat.f2py.rcm2points` before, but now we converted it to `rcm2points` plus an initial `import geocat.f2py`. That being said, I ran both pre-commit yapf and our older yapf (i.e. `python3 -m yapf --in-place --recursive --style google .` of the `pip3`-installed yapf), and both made the exact same changes as follows:

```
rcm2points(lat2d, lon2d, fi_nom.astype(np.float64), lat, lon,
           opt=2))
```

```
geocat.f2py.rcm2points(lat2d, 
                       lon2d, 
                       fi_nom.astype(np.float64), 
                       lat, 
                       lon,
                       opt=2))
```

It looks strange to me :)

**Review by pilotchute — APPROVED** [review]
